[jit] fix the NoneType param/buffer hack

suo opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack):
* **#32745 [jit] fix the NoneType param/buffer hack**

Some parameters (like `bias` in conv) are optional. To achieve this
previously, you had to add `bias` as a constant, which would invoke some
pretty weird behavior in the frontend, summarized as:
```
if bias is not None:
  add it as a parameter normally
else: # bias is None
  add it as a constant with the value None
```

There are several things bad about this:
1. Bias is not a constant. Marking it `__constants__` is confusing.
2. It basically relies on an implementation detail (the frontend
processes parameters before constants) to work.

Okay, whatever. I don't even know why we did this originally, but
getting rid of it doesn't break anything, so I assume improved NoneType
refinement has made this a non-issue.

Note on perf: this will make no difference; if bias was `None` it's still
folded out today, if bias is a Tensor it would be added as a parameter
both before and after this change

Differential Revision: [D19628634](https://our.internmc.facebook.com/intern/diff/D19628634)

suo committed:
[jit] fix the NoneType param/buffer hack

Some parameters (like `bias` in conv) are optional. To achieve this
previously, you had to add `bias` as a constant, which would invoke some
pretty weird behavior in the frontend, summarized as:
```
if bias is not None:
  add it as a parameter normally
else: # bias is None
  add it as a constant with the value None
```

There are several things bad about this:
1. Bias is not a constant. Marking it `__constants__` is confusing.
2. It basically relies on an implementation detail (the frontend
processes parameters before constants) to work.

Okay, whatever. I don't even know why we did this originally, but
getting rid of it doesn't break anything, so I assume improved NoneType
refinement has made this a non-issue.

Note on perf: this will make no difference; if bias was `None` it's still
folded out today, if bias is a Tensor it would be added as a parameter
both before and after this change

[ghstack-poisoned]

suo commented:
cc @jerryzh168 this should fix your issue

suo commented on torch/jit/_recursive.py:
```diff
@@ -140,13 +137,23 @@ def infer_type(name, item):
+            elif name in nn_module._parameters:
+                hint = "parameter"
+            else:
+                raise AssertionError("added_names must be submodule, parameter, or buffer")
```
Attributes are handled separately (and last), because unlike modules/buffers/parameters they don't have an actual definition in the nn.Module API, they are just Python attributes. The purpose of added_names is to make sure we don't re-add attributes that already exist as modules/buffers/parameters, so by definition it can't be an attr.

suo commented on torch/jit/_recursive.py:
```diff
@@ -87,27 +87,24 @@ def infer_type(name, item):
+            #
+            # So to achieve the nn.Module behavior, add the NoneType parameters
+            # as an attribute.
+            concrete_type_builder.add_attribute(name, attr_type, False)
```
`False` is whether it is a parameter or not. I can update the comment

zdevito replied:
also the only difference across the if statement is the boolean so:
```
concrete_type_builder.add_attribute(name, attr_type, item is not None)
```

zdevito reviewed (approved):
nice cleanup!

suo commented on torch/jit/_recursive.py:
```diff
@@ -87,27 +87,24 @@ def infer_type(name, item):
+        assert item is None or isinstance(item, torch.Tensor)
         attr_type = infer_type(name, item)
-        concrete_type_builder.add_attribute(name, attr_type, True)
+        if item is None:
```
This appears to work at least on the surface (the new `script::Module` API is smart enough to emulate the behavior of `nn.Module` under these conditions)

suo committed:
Update on "[jit] fix the NoneType param/buffer hack"


Some parameters (like `bias` in conv) are optional. To achieve this
previously, you had to add `bias` as a constant, which would invoke some
pretty weird behavior in the frontend, summarized as:
```
if bias is not None:
  add it as a parameter normally
else: # bias is None
  add it as a constant with the value None
```

There are several things bad about this:
1. Bias is not a constant. Marking it `__constants__` is confusing.
2. It basically relies on an implementation detail (the frontend
processes parameters before constants) to work.

Okay, whatever. I don't even know why we did this originally, but
getting rid of it doesn't break anything, so I assume improved NoneType
refinement has made this a non-issue.

Note on perf: this will make no difference; if bias was `None` it's still
folded out today, if bias is a Tensor it would be added as a parameter
both before and after this change

[ghstack-poisoned]

kostmo commented:
## :pill: CircleCI build failures summary and remediations
As of commit a9b667e9:


* **1/1** failures introduced in this PR


## Detailed failure analysis

One may explore the probable reasons each build failed interactively [on the Dr. CI website](https://dr.pytorch.org/commit-details.html?sha1=a9b667e90a2684fcd23a6f75b6ae4845f869ed1f).

### :detective: 1 new failure recognized by patterns
The following build failures do not appear to be due to upstream breakage:
#### [![See CircleCI build](https://avatars0.githubusercontent.com/ml/7?s=12)](https://circleci.com/gh/pytorch/pytorch/4373423) pytorch_macos_10_13_py3_test (1/1)
**Step:** "Test" ([full log](https://dr.pytorch.org/api/view-log-full?build_id=92802073) | [pattern match details](https://dr.pytorch.org/build-details.html?build_id=92802073))
<details>
<summary>
<code>Jan 29 14:03:02 RuntimeError: test_autograd failed!</code>
</summary>

```
Jan 29 14:03:02 Ran 1046 tests in 285.239s 
Jan 29 14:03:02  
Jan 29 14:03:02 FAILED (errors=1, skipped=13, expected failures=1) 
Jan 29 14:03:02  
Jan 29 14:03:02 Generating XML reports... 
Jan 29 14:03:02 Traceback (most recent call last): 
Jan 29 14:03:02   File "test/run_test.py", line 456, in <module> 
Jan 29 14:03:02     main() 
Jan 29 14:03:02   File "test/run_test.py", line 449, in main 
Jan 29 14:03:02     raise RuntimeError(message) 
Jan 29 14:03:02 RuntimeError: test_autograd failed! 
Jan 29 14:03:02 + cleanup 
Jan 29 14:03:02 + retcode=1 
Jan 29 14:03:02 + set +x 
```

</details>



---

<details><summary>This comment was automatically generated by <a href="https://github.com/kostmo/circleci-failure-tracker/tree/master/docs/from-pull-request-comment">Dr. CI</a> (expand for details).</summary>Follow <a href="https://dr.pytorch.org/admin/comments-opt-out.html">this link to opt-out</a> of these comments for your Pull Requests.

Please report bugs/suggestions on the <a href="https://github.com/kostmo/circleci-failure-tracker/issues">GitHub issue tracker</a>.
</details>
